include/ofi_mr.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2017 Intel Corporation, Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+
+#include "fi.h"
+#include "fi_util.h"
+
+/*
+ * MR cache
+ */
+
+struct ofi_mr_cache_attr {
+	int			reg_size;
+	int			stale_size;
+	int			elem_size;
+	struct fi_provider	*prov;
+	/*struct ofi_mr_monitor	*mr_monitor;*/
+	void			*reg_context;
+	void			*dereg_context;
+	void *(*reg_callback)(void *handle, void *address, size_t length,
+			      void *reg_context, void *context);
+	int (*dereg_callback)(void *handle, void *context);
+};
+
+struct ofi_mr_cache {
+	struct ofi_mr_cache_attr	attr;
+	/*struct ofi_mr_monitor_eq	eq;*/
+	struct dlist_entry		lru_list;
+	RbtHandle			inuse_tree;
+	uint64_t			inuse_elem;
+	RbtHandle			stale_tree;
+	uint64_t			stale_elem;
```

Comment:
- Use a single tree to manage registrations.
- Define a new struct ofi_mr_cache_entry
```
struct ofi_mr_cache_entry {
    dlist_entry entry; /* for lru_list */
    ofi_atomic32_t refcnt;
    char data[0]; /* aligned, elem_size data */
};
```

Reply:
This comment hasn't been applied yet.
I'm waiting when the #3569 PR gets merged. 


I'll continue to integrate MR monitor and re-implement integrationof MR caching to the verbs level.

This is needed to validate the applied changes for these comments.